### PR TITLE
Auto merge pull requests when all checks succeeded via WebUI

### LINKED ISSUE #19621: Auto merge pull requests when all checks succeeded via WebUI
to work use #19648 or rewrite ...

Bildschirmfoto zu 2021-04-10 15-00-20

Bildschirmfoto zu 2021-04-10 15-03-24

### PR DESCRIPTION
**Pull request by 6543**
taken from #9307, they where removed to get the feature in ...  
close #19621

TODO:
 - [x] backend
 - [x] frontend

Screenshot from 2022-05-20 03-37-10
Screenshot from 2022-05-20 03-37-34
Screenshot from 2022-05-20 03-41-22

### CODE REVIEW

**Comment by 6543** [comment]
Copy the comments here (some thoughts about how to refactor this UI):

----

Either rewrite the merge form with Vue.

Or, just add an checkbox `Auto merge when succeed` to merge UI, then these code do not need to be copied twice.

* When the PR is merge-able, hide the chekcbox, show merge buttons.
* When the PR is not merge-able, show the checkbox and hide merge buttons. Only when the checkbox is checked, then show the merge buttons.

(ps: the `checkbox` could be some other component, maybe switch between AutoMerge and ManualMerge)

File: routers/web/repo/pull.go
Comment (6543) [inline routers/web/repo/pull.go:1077]:
```suggestion
	if err := automerge.RemoveScheduledAutoMerge(ctx, ctx.Doer, issue.PullRequest); err != nil {
```
if #19649 got merged

File: routers/web/repo/pull.go
Comment (6543) [inline routers/web/repo/pull.go]:
```suggestion
	if err := automerge.RemoveScheduledAutoMerge(ctx, ctx.Doer, issue.PullRequest); err != nil {
		if db.IsErrNotExist(err) {
```

if #19649 got merged

**Comment by 6543** [comment]
@wxiaoguang would you mind to take over the vue part?

I invited you: https://github.com/6543-forks/gitea/invitations so you could directly push to this branch

**Comment by wxiaoguang** [comment]
UI demo.


image

image

image


image

**Comment by 6543** [comment]
```
/drone/src/web_src/js/components/PullRequestMergeForm.vue

  77:5   warning  Require self-closing on HTML elements (<div>)  vue/html-self-closing

  77:54  warning  'v-html' directive can lead to XSS attack      vue/no-v-html
```
@wxiaoguang

**Comment by wxiaoguang** [comment]
This is a draft demo, some details haven't been polished. If it can be confirmed to be acceptable, then it can be polished then.

**Comment by 6543** [comment]
nice idea like it :)

2 todos:
- [ ] if pull is already shedulled you still have the option to shedulle it again
- [ ] if you are no admin but have merge access you should also be able to shedulle/cancle it

**Comment by 6543** [comment]
well I think, **we should not have to document it**
it should be self explaining, by the UI to only show options that are posible...

- any user can create a pull to a repo as long as it's visibe to them
- codeWrite permission allow you to merge beside protected branch is protecting from merging
- auto-merge can be scheduled/cancled by codeWrite permission
- admins & owners can merge (with red button) with bypassing protected branch rules - but also are able to schedule/cancle auto-merge

**Comment by 6543** [comment]
```
templates/repo/issue/view_content/pull.tmpl:
	259: Wrong indentation type(spaces instead of tabs)
```

**Comment by 6543** [comment]
... I'm on it

File: web_src/less/_base.less
Comment (wxiaoguang) [inline web_src/less/_base.less:2013]: I guess some reviewers may ask "why" here. So I answer ahead: these code doesn't make sense, and the problem described by #10835 doesn't exist at the moment (as least for my testing).

There should be no global pollution for `max-width`, there are also more issues related to it, in which users complain about "why the width of the dropdown is so narrow".

File: web_src/js/components/PullRequestMergeForm.vue
Comment (wxiaoguang) [inline web_src/js/components/PullRequestMergeForm.vue]: Answer the `eslint-disable` ahead: at the moment `hasPendingPullRequestMergeTip` is generated as HTML content by template, so the `v-html` must be used here.

**Comment by 6543** [comment]
for writePermission:
this should not be red - I would suggest blue ...:
Screenshot from 2022-05-19 18-09-26

for admin/ownerPermission:
this should not have a "schedule automerge section if it's already sceduled" - if no schedule this is fine
Screenshot from 2022-05-19 18-08-40

**Comment by wxiaoguang** [comment]
> for writePermission: this should not be red - I would suggest blue ...: Screenshot from 2022-05-19 18-09-26

Updated

> for admin/ownerPermission: this should not have a "schedule automerge section if it's already sceduled" - if no schedule this is fine Screenshot from 2022-05-19 18-08-40

I think users should be able to change the merge style even if there is an existing schedule, so keeping is better for them.
And removing it would cause the UI code become more complex, for example, more hiding work for writers.

**Comment by 6543** [comment]
ok vue code for blue work but the css dont apply it (it stays green) :laughing: 
Screenshot from 2022-05-19 18-51-31

**Comment by wxiaoguang** [comment]
image


The blue button is overwritten by dirty `!important` pollution in dark theme.

**Review by 6543 — COMMENTED** [review]
LGTM for frontend :tada:

**Comment by 6543** [comment]
@wxiaoguang I think you can still lgtm at least the backend ;)

**Comment by wxiaoguang** [comment]
I think the backend might need to be changed a little (according to: "I think users should be able to change the merge style even if there is an existing schedule (instead of canceling it and re-submit the form), so keeping is better for them.")

At this moment , IIRC, the backend would return an error if the auto merge schedule is re-submitted again.

Could you take a look at it?

File: templates/repo/issue/view_content/pull.tmpl
Comment (6543) [inline templates/repo/issue/view_content/pull.tmpl:361]: 'hideAutoMerge': true ... should be always true if you can merge directly ...

**Comment by 6543** [comment]
@wxiaoguang it's a question about UX ...:

should you have to `cancle -> select new one` or `direcly` be able to overwrite ...

File: templates/repo/issue/view_content/pull.tmpl
Comment (wxiaoguang) [reply templates/repo/issue/view_content/pull.tmpl:361]: Updated.

**Comment by 6543** [comment]
I'll commit this in ~40min or so :)

could you look into https://github.com/go-gitea/gitea/pull/19648#discussion_r877361473 meanwhile?

**Review by wxiaoguang — APPROVED** [review]

**Comment by delvh** [comment]
How do I enable this UI?
On my setup, the branch looks like this: 
image

**Comment by 6543** [comment]
enable branch protection

**Comment by delvh** [comment]
> enable branch protection and make status required

Wait, to recap: It is possible to auto-merge PRs but **only if branch protection is enabled for the branch and the status is required**? That sounds too restrictive.
Why shouldn't it always be possible when branch protection is enabled for that branch?
Because this would for example allow to auto-merge a PR once it has enough approvals if there is no CI.
I think that is the most user-friendly approach...
Also, it aligns with the current definition (`merge when all checks succeed` - and enough approvals is technically a check.)

**Comment by 6543** [comment]
> enable branch protection and make status required

this was just an example to make sure you can test it ...

the other options you describe also work ...

**Comment by 6543** [comment]
one thing: recheck is only triggered by an status update atm ... but since this pull here is only for adding UI I think we could add other recheck triggers in other pulls @delvh

**Comment by 6543** [comment]
@delvh do we need something else?

**Comment by delvh** [comment]
I'll get to testing it once I find the time for it, I currently have far too little time.
Hopefully, I will still be able to test itthis weekend.

**Comment by 6543** [comment]
Thanks ❤️

**Review by delvh — CHANGES_REQUESTED** [review]
More things I noticed while testing:
1. I don't think the current design is the most user-friendly option: Users first have to click on the drop-down to even know that this feature exists.
What percentage of users do you think clicks (regularly) on that dropdown?
I doubt that number is larger than 40%.
This means with the current approach, the majority of users won't even know the feature exists for a really long time.
2. This is one of the few things that GitLab actually does nicely: 
image
with the corresponding repo settings:
image
If there is a pending pipeline and the setting is true, there is no way to merge the PR prematurely. Otherwise, if the pipeline has terminated, the user can only merge the branch directly.
That approach is actually quite user-friendly.
Unfortunately, Gitea currently has no CI, so that option isn't possible (for now).
3. But still, what about using i.e. a checkbox `only once checks succeed` next to the merge button instead?
4. General issue: with the current approach, the option to select another merge style seems not user-friendly, especially in regards to the automerge:
I don't exactly expect that clicking on any of the drop-down options only changes the merge style without "merging" directly. This can be noticed especially for the automerge, because it is not immediately clear that clicking on one of the dropdown-items does not apply the strategy directly.
5. There seems to be a CSS issue (I'm not hovering over anything):
image
Also, when that CSS bug occurs, it is impossible to not schedule an auto merged style (so i.e. it is only possible to select an auto-rebase and not a normal rebase).
If I had to guess, that issue occurs when an auto merge has been scheduled and the page is reloaded.
6. as can be seen in image as well: Why am I able to auto schedule multiple subsequent times without canceling the automerge first?
7. When you cancel an auto merge, the merge style still stays on the auto merge. That, combined with 5, is a bad combi as it completely disables returning back to a normal merge style.

**Comment by wxiaoguang** [comment]
`5 & 7. There seems to be a CSS issue (I'm not hovering over anything):`

It's by purpose , not a CSS issue or bug. In such case, the only option is `Auto merge`, so the hint is always shown.

`6. Why am I able to auto schedule multiple subsequent times without canceling the automerge first?`

If you want to change the commit message, why do you need 2 more clicks ? Why not update it directly?

----

Feel free to rewrite it.

**Comment by 6543** [comment]
> If there is a pending pipeline and the setting is true, there is no way to merge the PR prematurely. Otherwise, if the pipeline has terminated, the user can only merge the branch directly.
That approach is actually quite user-friendly.
Unfortunately, Gitea currently has no CI, so that option isn't possible (for now).

I'm against to build a CI into gitea!!! - If you think UX is not nice, we should define better ways how CIs can **integrate**
And the whole point of protected branches is, that  only admins can merge stuff - so if pipeline is pending it should NOT be able to merge

**Comment by 6543** [comment]
to 3., I find gitlab's ui to overloaded, you have 10 buttons just for the sake of throw them in the face of a user?!?

**Comment by 6543** [comment]
> 7. "...When you cancel an auto merge, the merge style still stays on the auto merge. ..."

well the other option is no merge at all in this stage - should we disable the merge button and make dropdown aktivate it?

File: templates/repo/issue/view_content/pull.tmpl
Comment (lunny) [inline templates/repo/issue/view_content/pull.tmpl]: Looks like the variable is not used?

**Comment by lunny** [comment]
I also think the UI/UX may make somebody confusing. I suggest we have more discuss about that. And I think the auto merge should also require the base branch is updated?

**Comment by 6543** [comment]
@lunny it does respect base branch must be up to date if you set it

**Comment by wxiaoguang** [comment]
> I also think the UI/UX may make somebody confusing. I suggest we have more discuss about that. And I think the auto merge should also require the base branch is updated?

Please propose non-confusing UI/UX

**Comment by wxiaoguang** [comment]
I do not think that an admin should have to cancel an scheduled auto merge before they do a normal merge. The admin should be able to bypass the checks and scheduled auto merge and do the normal merge in one click.

But if you prefer to change the logic, there is no blocker from my side, edit on this PR directly.

**Comment by lunny** [comment]
> I do not think that an admin should have to cancel an scheduled auto merge before they do a normal merge. The admin should be able to bypass the checks and scheduled auto merge, then do the normal merge in one click.
> 
> But if you prefer to change the logic, there is no blocker from my side, edit on this PR directly.

You are right. Admin should not be blocked by that. But I think that's what the complicated thing. We have to put both manually merge and auto merge in the UI in that situation and have to have a lock between manual merge and auto merge. To avoid that, we can simplify the process.

I'm not sure if this is the right direction. Just my thought and more ideas are welcome.

**Comment by wxiaoguang** [comment]
About this PR, I think it should either:

* get merged and improved later.
* (OR) get improved before 1.17-rc and get merged in 1.17


Otherwise these already merged backend code and translations would become dead code and might be more difficult to be maintained in the future.

**Comment by wxiaoguang** [comment]
@techknowlogick Do you prefer to move it to 1.18? Can you help me to know the reasons?

**Comment by wxiaoguang** [comment]
Some UI adjustments (shorten the texts)

image

**Comment by lunny** [comment]
I think the merge style should be decided when starting to auto merge. The change of merge style should not be allowed before cancel the auto merge.

File: web_src/js/components/PullRequestMergeForm.vue
Comment (delvh) [inline web_src/js/components/PullRequestMergeForm.vue]: Aren't these parentheses now duplicated because it's already in the translated string?

Comment (delvh) [reply web_src/js/components/PullRequestMergeForm.vue]: As I expected:
image

**Review by delvh — APPROVED** [review]
Tested it.
Seems to work now as intended (apart from the small translation issue).
Let's keep improving the UI for another PR and get this merged for 1.17.

**Comment by 6543** [comment]
:rocket:

**Pull request merged**
